src/Microsoft.AspNetCore.Server.Kestrel.Core/Internal/Http/MessageBody.cs

```diff
@@ -273,6 +274,22 @@ private void TryStartTimingReads()
             }
         }
 
+        private void TryPauseTimingReads()
+        {
+            if (!RequestUpgrade)
```

Comment:
Reads return 0 bytes on upgrade requests if you try to read from them before upgrading. So the app wouldn't hang on a read. Is that your concern?